compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -342,6 +342,12 @@ impl<'a> Parser<'a> {
         let span = lo.to(self.prev_token.span);
         let mut ty = self.mk_ty(span, kind);
 
+        if self.eat_keyword(sym::is) {
```

Comment:
doesn't this break macro code like:

```
macro_rules! foo {
  ($ty:ty) => {};
  (() is $pat:pat) => {};
}

foo!(() is 0..1);
```

Reply:
macro rules matches continue being the least forwards compatible thing ever :|

Reply:
Maybe it's better to expose this through a builtin macro instead of surface syntax, like `rustc_pattern_type!()` which expands into ast::TyKind::Pat

Reply:
Or to `builtin#PatTy(…)` (with `builtin#` from [T-compiler MCP #580](https://github.com/rust-lang/compiler-team/issues/580)).

Reply:
> macro rules matches continue being the least forwards compatible thing ever :|

And this is where the editions system can shine.

Reply:
... I'm a little surprised/sad that `foo!` works at all today. Follow sets and greedy parsing were supposed to handle this -- you're not allowed to follow a `:ty` fragment with `is` -- but follow set restrictions are only applied within an arm, and not across arms. Being able to follow a fragment with some token in a simultaneously-taken arm matching the fragment's tokens essentially makes follow sets ineffectual at their job of making fragments forwards compatible. But simultaneously, applying follow set rules between arms would be very restrictive and probably not even solve the problem entirely, since you can attack non-follow-set grammar extensions similarly. I guess you'd need to ensure that the choice of which arm to take is always known to be forwards-compatibly LL(k) via some sort of "leading set," and if the choice isn't LL(k), ensure that the refining match fits within the grammar and follow set of the refined fragment.

i.e., complicated. And still overly restrictive, since we also have the rule for lexically prior arms taking priority over latter arms, so there's only a forwards compatibility concern if the refining arms come after the refined arm. (And until today I thought that macro parsing was greedier than it actually is, preventing `foo!` from ever matching the second arm here, since the `$:ty` arm is prior and matched the prefix. Despite probably relying on that not being the case.)